compiler-lib/src/sem.rs

```diff
@@ -1,6 +1,163 @@
-use crate::{ast, context};
-use failure::Error;
+use crate::{asciifile::Spanned, ast, context, strtab::Symbol, visitor::NodeKind};
+use failure::{format_err, Error, ResultExt};
+use std::{collections::HashMap, rc::Rc};
 
-pub fn check<'a, 'f>(_ast: &'a ast::AST<'f>, _context: &context::Context<'_>) -> Result<(), Error> {
-    unimplemented!();
+type ClassesAndMembers<'a, 'f> = HashMap<
+    (Symbol, Symbol, ast::ClassMemberKindDiscriminants),
+    &'a Spanned<'f, ast::ClassMember<'f>>,
+>;
+
+pub fn check<'a, 'f>(ast: &'a ast::AST<'f>, context: &context::Context<'_>) -> Result<(), Error> {
+    let classes_and_members = ClassesAndMembersVisitor::visit(ast, context)?;
+
+    // Let's draw some pretty annotations that show we classified classes + their
+    // members and method correctly
+    for ((classsym, membersym, memberytypediscr), decl) in &classes_and_members {
+        // FIXME: Need ClassMember.name be Spanned instead of Symbol
+        let highlightspan = &decl.span;
+        context.diagnostics.info(&Spanned {
+            span: highlightspan.clone(),
+            data: format!("{}.{} ({})", classsym, membersym, memberytypediscr),
+        });
+    }
+
+    // Now the second pass, this one will do (in one traversal)
+    //  - name analysis of the AST
+    //  - type checking of expressions
+    SecondPass::visit(ast, context, &classes_and_members)
+}
+
+struct ClassesAndMembersVisitor;
+
+impl ClassesAndMembersVisitor {
+    pub fn visit<'a, 'f>(
+        ast: &'a ast::AST<'f>,
+        context: &context::Context<'_>,
+    ) -> Result<ClassesAndMembers<'a, 'f>, Error> {
+        let mut res = HashMap::new();
+        Self::do_visit(&NodeKind::from(ast), context, &mut res)?;
+        Ok(res)
+    }
+
+    fn do_visit<'a, 'f>(
+        n: &NodeKind<'a, 'f>,
+        _context: &context::Context<'_>,
+        res: &mut ClassesAndMembers<'a, 'f>,
+    ) -> Result<(), Error> {
+        use self::{ast, NodeKind::*};
+        match n {
+            AST(ast::AST::Program(p)) => {
+                for class in &p.data.classes {
+                    for member in class.members.iter() {
+                        let discr = ast::ClassMemberKindDiscriminants::from(&member.kind);
+                        res.insert(
+                            (Rc::clone(&class.name), Rc::clone(&member.name), discr),
+                            &member,
+                        );
+                    }
+                }
+                Ok(())
+            }
+            x => {
+                return Err(format_err!(
```

Comment:
why not tighten the type definition of do_visit to ast::Ast::Program() to get rid of this impossible state?

Reply:
Or the only other possible state seems to be `ast::AST::Empty`, why not just reject the program at this point with "missing main"?

Reply:
There could be a lot of other states. `NodeKind` could be any `AST`-node. So if you call this function with an `Expr` it will error with
```
"unexpected ast node passed to first pass: Expr",
```

Reply:
I don't get it. He is not using recursion at all, since all ClassDefintions are at the top level. So why make the method parameter a AST-Node and not the Root Node Program/Empty which avoids this issue?

Reply:
Yeah this doesn't really need a visitor in that case then.